src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/PossibleActionsRuleEngine.java

```diff
@@ -0,0 +1,279 @@
+/**
+ * The MIT License
+ * Copyright (c) 2018 Estonian Information System Authority (RIA),
+ * Nordic Institute for Interoperability Solutions (NIIS), Population Register Centre (VRK)
+ * Copyright (c) 2015-2017 Estonian Information System Authority (RIA), Population Register Centre (VRK)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.niis.xroad.restapi.service;
+
+import ee.ria.xroad.signer.protocol.dto.CertRequestInfo;
+import ee.ria.xroad.signer.protocol.dto.CertificateInfo;
+import ee.ria.xroad.signer.protocol.dto.KeyInfo;
+import ee.ria.xroad.signer.protocol.dto.KeyUsageInfo;
+import ee.ria.xroad.signer.protocol.dto.TokenInfo;
+
+import org.springframework.stereotype.Component;
+
+import java.util.EnumSet;
+
+/**
+ * Validation logic for possible actions done on tokens, keys, certs and csrs
+ */
+@Component
+public class PossibleActionsRuleEngine {
+
+    // duplicate definition, since we dont want add direct dependency on signer
+    public static final String SOFTWARE_TOKEN_ID = "0";
+
+    /**
+     * Get possible actions for a token
+     */
+    public EnumSet<PossibleActionEnum> getPossibleTokenActions(TokenInfo tokenInfo) {
+        EnumSet<PossibleActionEnum> actions = EnumSet.noneOf(PossibleActionEnum.class);
+
+        if (tokenInfo.isActive()) {
+            actions.add(PossibleActionEnum.GENERATE_KEY);
+        }
+
+        if (tokenInfo.isActive()) {
+            actions.add(PossibleActionEnum.TOKEN_DEACTIVATE);
```

Comment:
It could, but I think I would prefer to keep it separate. I think it helps to be able to look at each PossibleActionEnum value individually and see clearly all the conditions that cause one individual action to be enabled. And that maybe makes it a bit easier to take one action and compare it's activation to the old codebase.

This sounds a bit silly since those two tokenInfo.isActive's are so simple and close to each other but that was kind of the thought behind it, it was not completely coincidental. The principle is maybe more clear in `getPossibleCertificateActions` and `getPossibleKeyActions`, where I am sure the if-blocks could be optimized, combined and simplified somewhat by combining actions, but keeping them separated is better. And using the same approach for `getPossibleTokenActions` is nice, IMO.